src/software/ai/hl/stp/tactic/goalie/goalie_fsm.h

```diff
@@ -0,0 +1,448 @@
+                         Line(goalie_restricted_area.posXPosYCorner(),
+                              goalie_restricted_area.posXNegYCorner()));
+        auto pos_side_x_goal =
+            intersection(Line(goalie_desired_position, field.friendlyGoalCenter()),
+                         Line(goalie_restricted_area.posXPosYCorner(),
+                              goalie_restricted_area.negXPosYCorner()));
+        auto neg_side_x_goal =
+            intersection(Line(goalie_desired_position, field.friendlyGoalCenter()),
+                         Line(goalie_restricted_area.posXNegYCorner(),
+                              goalie_restricted_area.negXNegYCorner()));
+
+        // if the goalie restricted area already contains the point, then we are
+        // safe to move there.
+        if (contains(goalie_restricted_area, goalie_desired_position))
+        {
+            return std::make_optional<Point>(goalie_desired_position);
+        }
+        // Due to the nature of the line intersection, its important to make sure the
+        // corners are included, if the goalies desired position intersects with width
+        // (see above), use those positions The last comparison is for the edge case when
+        // the ball is behind the net
+        else if (width_x_goal &&
+                 width_x_goal->y() <= goalie_restricted_area.posXPosYCorner().y() &&
+                 width_x_goal->y() >= goalie_restricted_area.posXNegYCorner().y() &&
+                 field.friendlyGoalCenter().x() <= goalie_desired_position.x())
+        {
+            return std::make_optional<Point>(*width_x_goal);
+        }
+
+        // if either two sides of the goal are intercepted, then use those positions
+        else if (pos_side_x_goal &&
+                 pos_side_x_goal->x() <= goalie_restricted_area.posXPosYCorner().x() &&
+                 pos_side_x_goal->x() >= goalie_restricted_area.negXPosYCorner().x())
+        {
+            return std::make_optional<Point>(*pos_side_x_goal);
+        }
+        else if (neg_side_x_goal &&
+                 neg_side_x_goal->x() <= goalie_restricted_area.posXNegYCorner().x() &&
+                 neg_side_x_goal->x() >= goalie_restricted_area.negXNegYCorner().x())
+        {
+            return std::make_optional<Point>(*neg_side_x_goal);
+        }
+
+        // if there are no intersections (ex. ball behind net), then we are out of luck
+        else
+        {
+            return std::nullopt;
+        }
+    }
+
+    /**
+     * Gets the position for the goalie to move to, to best position itself between the
+     * ball and the friendly goal
+     * @param ball the ball to position the goalie relative to
+     * @param field the field to position the goalie on
+     * @param goalie_tactic_config the goalie tactic config
+     *
+     * @return the position that the goalie should move to
+     */
+    static Point getGoaliePositionToBlock(
```

Comment:
What do you think about using calcbestshotongoal to position the goalie? With #2030, the crease defenders would be independent of the goalie position.

Reply:
Sure, seems like a good idea - when you say the crease defenders would be independent of the goalie, how did this affect the goalie before? i thought the goalie affected the crease defenders' positions, but the goalie was always independent, so I'm a bit confused where that comes into play here

Reply:
Also, to clarify, this would be very similar to how panic state works, where the ball velocity in panic_s would be the best shot on goal in position_to_block_s, then find the closest point to the line from the ball to the target of the best shot?

Reply:
Yeah just an idea. I think we'll discuss this after this PR, crease defender, and shadow enemy are done